libavcodec/rkmppdec.c

```diff
@@ -516,24 +518,43 @@ static int ffrkmpp_receive_frame(AVCodecContext *avctx, AVFrame *frame)
 
 static int ffrkmpp_decode(AVCodecContext *avctx, void *frame, int *got_frame, AVPacket *avpkt)
 {
+    int buffer_full = 0;
+    int decoder_timeout = 0;
+    int retries = 0;
     int ret;
 
     *got_frame = 0;
 
-    if (avpkt) {
-        if (avpkt->data == NULL && avpkt->size == 0)
-            ret = ffrkmpp_send_packet(avctx, NULL);
-        else
-            ret = ffrkmpp_send_packet(avctx, avpkt);
-        if (ret < 0 && ret != AVERROR(EAGAIN) && ret != AVERROR_EOF)
+    do {
+        if (avpkt) {
+            if (avpkt->data == NULL && avpkt->size == 0)
+                ret = ffrkmpp_send_packet(avctx, NULL);
+            else
+                ret = ffrkmpp_send_packet(avctx, avpkt);
+            if (ret < 0 && ret != AVERROR(EAGAIN) && ret != AVERROR_EOF) {
+                av_log(avctx, AV_LOG_ERROR, "Failed to send packet (code = %d)\n", ret);
+                return ret;
+            }
+            buffer_full = ret == AVERROR(EAGAIN);
+        }
+
+        ret = ffrkmpp_receive_frame(avctx, frame);
+        if (ret < 0 && ret != AVERROR(EAGAIN) && ret != AVERROR_EOF) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to receive frame (code = %d)\n", ret);
             return ret;
+        }
+        if (ret >= 0)
+            *got_frame = 1;
+
+        decoder_timeout = ret == AVERROR(EAGAIN);
     }
+    while (buffer_full && decoder_timeout && retries++ < 5);
 
```

Comment:
I am wondering if that is the rigth way to handle that situation. I cannot recall any hwdec doing this.
If the issue could be the too short timeout maybe it would be better to make the timeout longer ?
Eventually then the best way would be to add a parameter to the hwdec that can be used to set a longer timeout ?